Source/WebCore/Modules/mediasource/MediaSourceHandle.cpp

```diff
@@ -0,0 +1,131 @@
+
+#include "MediaSource.h"
+#include "MediaSourcePrivate.h"
+#include "MediaSourcePrivateClient.h"
+
+namespace WebCore {
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(MediaSourceHandle);
+
+class MediaSourceHandle::Dispatcher final : public ThreadSafeRefCounted<Dispatcher> {
+public:
+    static Ref<Dispatcher> create(Function<void(Function<void()>&&, bool)>&& dispatcher) { return adoptRef(*new Dispatcher(WTFMove(dispatcher))); }
+
+    void dispatch(Function<void()>&& task, bool forceRunInWorker = false) const
+    {
+        m_dispatcher(WTFMove(task), forceRunInWorker);
+    }
+private:
+    Dispatcher(Function<void(Function<void()>&&, bool)>&& dispatcher)
+        : m_dispatcher(WTFMove(dispatcher))
+    {
+    }
+    const Function<void(Function<void()>&&, bool)> m_dispatcher;
+};
+
+Ref<MediaSourceHandle> MediaSourceHandle::create(MediaSource& mediaSource, Function<void(Function<void()>&&, bool)>&& dispatcher)
+{
+    return adoptRef(*new MediaSourceHandle(mediaSource, WTFMove(dispatcher)));
+}
+
+Ref<MediaSourceHandle> MediaSourceHandle::create(Ref<MediaSourceHandle>&& other)
+{
+    other->setDetached(false);
+    return other;
+}
+
+MediaSourceHandle::MediaSourceHandle(MediaSource& mediaSource, Function<void(Function<void()>&&, bool)>&& dispatcher)
+    : m_client(mediaSource.client().get())
+    , m_mediaSource(mediaSource)
+    , m_isManaged(mediaSource.isManaged())
+    , m_dispatcher(Dispatcher::create(WTFMove(dispatcher)))
+{
+}
+
+MediaSourceHandle::MediaSourceHandle(MediaSourceHandle& other)
+    : m_client(other.m_client)
+    , m_mediaSource(other.m_mediaSource)
+    , m_hasEverBeenAssignedAsSrcObject(other.m_hasEverBeenAssignedAsSrcObject)
+    , m_detached(false)
+    , m_isManaged(other.isManaged())
+    , m_dispatcher(other.m_dispatcher)
+{
+    ASSERT(!other.m_detached);
+    other.m_detached = true;
+    registerWithMediaSource();
+}
+
+MediaSourceHandle::~MediaSourceHandle() = default;
+
+void MediaSourceHandle::ensureOnDispatcher(Function<void()>&& task) const
+{
+    m_dispatcher->dispatch(WTFMove(task));
+}
+
+Ref<DetachedMediaSourceHandle> MediaSourceHandle::detach()
+{
+    return adoptRef(*new MediaSourceHandle(*this));
+}
+
+WeakPtr<MediaSource> MediaSourceHandle::mediaSource() const
+{
+    return m_mediaSource;
+}
+
+void MediaSourceHandle::registerWithMediaSource()
+{
+    m_dispatcher->dispatch([weakMediaSource = m_mediaSource, protectedThis = Ref { *this }]() mutable {
+        if (RefPtr mediaSource = weakMediaSource.get())
+            mediaSource->registerTransferredHandle(protectedThis);
+    }, true);
+}
+
+RefPtr<MediaSourcePrivateClient> MediaSourceHandle::mediaSourcePrivateClient() const
+{
+    return m_client.get();
+}
+
+void MediaSourceHandle::mediaSourceDidOpen(MediaSourcePrivate& privateMediaSource)
+{
+    Locker locker { m_lock };
+    m_private = privateMediaSource;
+}
+
+RefPtr<MediaSourcePrivate> MediaSourceHandle::mediaSourcePrivate() const
+{
+    Locker locker { m_lock };
```

Comment:
I would need to use the MediaSourceInterfaceProxy interface to pass the MediaSourcePrivate to the MediaSource rather than from the MediaSourcePrivateClient.

I'll take this as a follow-up